### PR TITLE
Fixes #5607 - Assets.json options

### LINKED ISSUE #5607: Assets.json: Differentiate between watch and build files
I'm not sure this will apply to everyone, but we have a main .less file that imports all other less files into it (a master.less, if you will).  With Assets.json, are you specifying the files to watch as well as compile in a single array?  For us, we would want to watch *.less (and **/*.less) and only build master.less.
Is it possible to split up the watch vs build?
Does that make sense?

### PR DESCRIPTION
**Pull request by dcinzona**
Added support for separation of watch vs build .less files.
This required adding a flag for forcing rebuild, because the watch files may have changed, but the build files may not have, causing rebuild to fail.  Setting the rebuildAlways flag to true forces the rebuild to occur, regardless of files changed.  
Also added an option to exclude sourcemaps.  This defaults to false (include them if not set).

All options are ... wait for it ... optional.

Example Assets.json:
```
[
  {
    "inputs": [ "Less/master.less" ],
    "output": "Styles/@.css",
    "watch": ["Less/**/*.less", "Less/*.less"],
    "rebuildAlways": true,
    "excludeSourceMaps": true
  }
]
```

### CODE REVIEW

**Comment by DanielStolt** [comment]
I really like what you did here. I have a couple of suggestions for improvement:

1. Why not get rid of the `rebuildAlways` parameter, and in the `watch` task always call `createAssetGroupTask()` with `doRebuild = true` same as the `rebuild` task does? I should have done that from the start, I think I just forgot.
2. The optional `watch` property you introduced should be treated as files to watch **in addition** to the inputs, not instead of the inputs IMO.
3. The `excludeSourcemaps` option is fine but I think we should switch it around so the property name is `generateSourceMaps` and it defaults to `true`.

**Comment by dcinzona** [comment]
Agree and will make the edits when I get into the office (it's 4:45am here :)). 
Regarding 1: if rebuild always is set in the watch task, why use newer at all?
2 is interesting, I'll have to play around with that. The reason this worked for me was that the watch list also includes the inputs file, but I can see where the case might be that it doesn't.  
3. Not a problem, and totally makes sense. Double negatives can be confusing

**Comment by DanielStolt** [comment]
Did my Github notification wake you up?? :)

1. Because the `build` task uses it, and will continue to use it.

**Comment by dcinzona** [comment]
Ha no, I was already awake (back problems).  I'm going to start working on this now.

**Comment by dcinzona** [comment]
Updated the PR with the changes listed in https://github.com/dcinzona/Orchard/commit/0fcf8c740978dfeb17f3ebb05f57e61db363b1c7

I was thinking that it would be nice to also watch Assets.json, and when that changes, to restart the whole watch task (if running). What do you think? I could do a separate PR or implement it into this one.

**Comment by DanielStolt** [comment]
That would be a very nice thing, yes. Separate PR please.

**Comment by dcinzona** [comment]
Sounds good, will do.

**Comment by DanielStolt** [comment]
I can't see that you updated the PR with the changes? #3 is addressed, but not #1 and #2?

**Comment by dcinzona** [comment]
2 is in there - see line 96 on the change. 
```
.concat(assetGroup.inputPaths); //include inputs in watch list
```
For number 1, did you want me to remove the variable altogether or include it as optional?  I included it as optional, but defaulting to true.

**Comment by DanielStolt** [comment]
Ah, I see 2 now. I expected to find it in the ´watch` task function rather than the `resolveAssetGroupPaths` function. I think it would be more semantically correct to combine the two lists in the ´watch` task function.

Re 1, yes by "get rid of the rebuildAlways parameter" I meant actually removing it. :)

A general thing: rather than doing `variable != undefined`  I think it's more robust to do `!!variable` as it handles null, empty string and undefined - right?

**Comment by dcinzona** [comment]
Yup, that works.  I'll update accordingly!

**Comment by dcinzona** [comment]
So, `!!` will return true if the variable value is set to `false`...
```javascript
    var generateSourceMaps = !!assetGroup.generateSourceMaps ? assetGroup.generateSourceMaps : true;
    console.log("source maps: " + generateSourceMaps); // returns true even if explicitly set to false
```
```javascript
    var generateSourceMaps = assetGroup.generateSourceMaps !== undefined ? assetGroup.generateSourceMaps : true;
    console.log("source maps: " + generateSourceMaps); //returns false when explicitly set, otherwise true
```

**Comment by DanielStolt** [comment]
Yes in that particular case, since the default should be `true` you can't use that construct.

**Comment by dcinzona** [comment]
Ok, I think it's good to go now... let me know if you notice any issues.  I switched the `!==` condition to `hasOwnProperty` in the last commit.  May help with legibility.

**Comment by dcinzona** [comment]
FYI, I have it working where if an Asset.json file changes, it reloads the watchers.  You mentioned you wanted this as a separate PR, so that's what I'm going to do.  PR https://github.com/OrchardCMS/Orchard/pull/5613

File: src/Gulpfile.js [hunk 1]
```diff
@@ -13,10 +13,28 @@ var glob = require("glob"),
...
+   Saved to /Modules/My.Custom.Module/Assets.json
+   [
+      {
+        "inputs": [ "Less/master.less" ], //Specifies which files to process during the Build task
+        "output": "Styles/@.css", //When @ is specified, each file specified in "inputs" will be converted to [filename].css
+        "watch": ["Less/*.less"], //specifies which files to watch, use this when you have a master Less file that imports other Less files
+        "generateSourceMaps": true //Will include source maps in unminified version, defaults to 'true'
+      }
+    ]
+    "inputs" and "output" are required.  All other properties are optional.
```
Comment (DanielStolt) [inline src/Gulpfile.js]: This comment should go. We're going to write documentation for this, and also include a commented sample `Assets.json` in module scaffolding code-gen.

File: src/Gulpfile.js [hunk 2]
```diff
@@ -13,10 +13,28 @@ var glob = require("glob"),
 	uglify = require("gulp-uglify"),
 	rename = require("gulp-rename"),
     concat = require("gulp-concat"),
-	header = require("gulp-header")
+	header = require("gulp-header"),
+    notify = require("gulp-notify")
```
Comment (DanielStolt) [inline src/Gulpfile.js]: Leave out gulp-notify.

File: src/Gulpfile.js [hunk 3]
```diff
@@ -69,6 +88,12 @@ function resolveAssetGroupPaths(assetGroup, assetManifestPath) {
     assetGroup.inputPaths = assetGroup.inputs.map(function (inputPath) {
         return path.join(assetGroup.basePath, inputPath);
     });
+    
+    assetGroup.watchPaths = !!assetGroup.watch ?
+        assetGroup.watch.map(function (watchPath) {
+            return path.join(assetGroup.basePath, watchPath);
+        }).concat(assetGroup.inputPaths) : assetGroup.inputPaths;
```
Comment (DanielStolt) [inline src/Gulpfile.js]: I still think this is the wrong place for this - the merging of input paths with watch paths should go in the "watch" task function, not here. The `watchPaths` property should retain its meaning and value, not be "magically" altered here to mean something different.

File: src/Gulpfile.js [hunk 4]
```diff
@@ -95,6 +120,13 @@ function buildCssPipeline(assetGroup, doRebuild) {
             throw "Input file '" + inputPath + "' is not of a valid type for output file '" + assetGroup.outputPath + "'.";
     });
     var doConcat = path.basename(assetGroup.outputFileName, ".css") !== "@";
+    if (!doRebuild) {
+        console.log("CSS will only rebuild if less files specified in 'inputs' are newer.");
+    }
+    else {
+        console.log("Force Rebuild is enabled, rebuilding all input files.");
+    }
```
Comment (DanielStolt) [inline src/Gulpfile.js]: Don't understand why you've added this logging. Seems arbitrary - shouldn't be here.

File: src/Gulpfile.js [hunk 5]
```diff
@@ -115,13 +147,14 @@ function buildCssPipeline(assetGroup, doRebuild) {
...
-        .pipe(sourcemaps.write())
+        .pipe(gulpif(generateSourceMaps, sourcemaps.write()))
         .pipe(gulp.dest(assetGroup.outputDir))
         .pipe(minify())
         .pipe(rename({
             suffix: ".min"
         }))
-        .pipe(gulp.dest(assetGroup.outputDir));
+        .pipe(gulp.dest(assetGroup.outputDir))
+        .pipe(gulpif(doRebuild, notify("Rebuild complete"), notify("Build process complete")));
```
Comment (DanielStolt) [inline src/Gulpfile.js]: Remove notification.

File: src/Gulpfile.js (hunk 3 above)
Comment (dcinzona) [inline src/Gulpfile.js]: I put it in the "resolveAssetGroupPaths, as that is where you were generating the input paths, but I will move it to the watch function.

**Comment by dcinzona** [comment]
Ok, I've updated patch-5.  Let me know if this works.  I'm still a little unclear as to what you are looking for regarding the watchPaths resolve code, but I moved it to the watch task generation method.  Hopefully that's what you were looking for.  If not, let me know.

**Comment by DanielStolt** [comment]
I probably didn't manage to fully convey what I was after, but I merged this PR now and applied a small refactoring on top of it. Thanks for this contribution @dcinzona - makes the Gulp pipeline even better!

**Pull request merged**